source/common/stats/thread_local_store.cc

```diff
@@ -43,6 +43,9 @@ ThreadLocalStoreImpl::~ThreadLocalStoreImpl() {
   ASSERT(shutting_down_ || !threading_ever_initialized_);
   default_scope_.reset();
   ASSERT(scopes_.empty());
+  ASSERT(scopes_to_cleanup_.empty());
```

Comment:
In CI this is where you are crashing, on this test:
```
TEST(ThreadLocalStoreThreadTest, ConstructDestruct) {
  SymbolTableImpl symbol_table;
  Api::ApiPtr api = Api::createApiForTest();
  Event::DispatcherPtr dispatcher = api->allocateDispatcher("test_thread");
  NiceMock<ThreadLocal::MockInstance> tls;
  AllocatorImpl alloc(symbol_table);
  ThreadLocalStoreImpl store(alloc);

  store.initializeThreading(*dispatcher, tls);
  { ScopePtr scope1 = store.createScope("scope1."); }
  store.shutdownThreading();
}
```
I think this is because prior to your change we could have the central cache entries outlive the ThreadLocalStore object and now we can't.

I think you might be able to resolve by running something like this from shutdownThreading():
```
    scope_ids = std::move(scopes_to_cleanup_);
    tls_cache_->runOnAllThreads(
        [scope_ids](OptRef<TlsCache> tls_cache) { tls_cache->eraseScopes(*scope_ids); });
```
with appropriate locks and decls.

Reply:
Ah, that makes total sense. I was still in the process of recreating locally for myself, but this will save me some time for sure.

Reply:
local repro should still be the first thing you do: I'd try something like:
```
bazel test --config=clang-asan test/common/stats:thread_local_store_test --runs_per_test=100
```
I'll bet that repros and then you can ensure your fix worked.

Reply:
or `--config=clang-tsan`